### PR TITLE
[4.x] Don't hang on a 103 response

### PR DESCRIPTION
**Pull request by yschimke**
Race condition means that a 103 might be considered to complete a request, such that the the actual headers are considered trailers.

#fixes https://github.com/square/okhttp/issues/7936

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt [hunk 1]
```diff
@@ -283,7 +285,10 @@ class Http2Stream internal constructor(
     val open: Boolean
     synchronized(this) {
       if (!hasResponseHeaders || !inFinished) {
-        hasResponseHeaders = true
+        val status = headers[Header.RESPONSE_STATUS_UTF8]?.toIntOrNull()
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt:288]: Do you mind retargeting `master`, then cherry-picking this back to okhttp_4x ?

I’m a little surprised to see status codes handled at this layer

File: okhttp/src/test/java/okhttp3/InformationalResponseCodeTest.kt [hunk 2]
```diff
@@ -30,19 +30,22 @@ class InformationalResponseCodeTest {
     recordFrames = true
   }
 
-  private var client = clientTestRule.newClient()
+  private var client = clientTestRule.newClientBuilder()
+    .followRedirects(false)
+    .build()
 
   @Test
   fun test103() {
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/InformationalResponseCodeTest.kt:38]: Does master have this test?

**Comment by yschimke** [comment]
I'll check and update this weekend.

File: okhttp/src/test/java/okhttp3/InformationalResponseCodeTest.kt (hunk 2 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/InformationalResponseCodeTest.kt:38]: No, this is a manual test, see line 24.  The problem is mockwebserver in 4.x doesn't have support for informational responses.  So tests on master are better.

File: okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt (hunk 1 above)
Comment (yschimke) [reply okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt:288]: I think it went here so I could make the minimal change.

I can try again to apply some of the changes directly, it's been a while, so it's a bit vague why it broken down.

Comment (swankjesse) [reply okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt:288]: I’m still not a fan of handling status codes at the protocol layer. Revert this part and keep the fix for the infinite wait?

Comment (yschimke) [reply okhttp/src/main/kotlin/okhttp3/internal/http2/Http2Stream.kt:288]: This feels unfortunate.  I really would like to handle this properly in 4.12.

**Pull request closed without merging**